java/integTest/src/test/java/glide/ConnectionTests.java

```diff
@@ -202,4 +243,74 @@ public void test_az_affinity_non_existing_az() {
         assertEquals(4, matchingEntries);
         azTestClient.close();
     }
+
+    @SneakyThrows
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void test_connection_timeout(boolean clusterMode) {
+        var backoffStrategy =
+                BackoffStrategy.builder().exponentBase(2).factor(100).numOfRetries(1).build();
+        var client = createConnectionTimeoutClient(clusterMode, 250, 20000, backoffStrategy);
+
+        // Run a long-running DEBUG SLEEP command to simulate a blocking operation
+        CompletableFuture<Void> runDebugSleep =
+                CompletableFuture.runAsync(
+                        () -> {
+                            try {
+                                if (client instanceof GlideClusterClient) {
+                                    ((GlideClusterClient) client)
+                                            .customCommand(new String[] {"DEBUG", "sleep", "7"}, ALL_NODES)
+                                            .get();
+                                } else if (client instanceof GlideClient) {
+                                    ((GlideClient) client).customCommand(new String[] {"DEBUG", "sleep", "7"}).get();
+                                }
+                            } catch (InterruptedException | ExecutionException e) {
+                                throw new RuntimeException("Error during DEBUG SLEEP command", e);
+                            }
+                        });
+
+        // Test case 1: Client connection failure due to timeout
+        CompletableFuture<Void> failToConnect =
+                CompletableFuture.runAsync(
```

Comment:
Even though futures in java are lazy, `runAsync` still may start execution of the future before you call `get` or `join`. It isn't reliable to depend on that lazyness.
Consider creating 3 `Runnable`s and run them altogether.

Reply:
mmm ok